src/CubicSpline.cpp

```diff
@@ -0,0 +1,755 @@
+}
+
+void CubicSpline::solve_straight()
+{
+	if(check_minimum_size()) { return; }
+	size_t last= m_points.size();
+	vector<float> results(m_points.size());
+	vector<float> diagonals(m_points.size());
+	solution_cache.solve_diagonals_straight(diagonals);
+	results[0]= 3 * (m_points[1].a - m_points[0].a);
+	prep_inner(last, results);
+	results[last-1]= 3 * loop_space_difference(
+		m_points[last-1].a, m_points[last-2].a, m_spatial_extent);
+
+	// The system of equations to be solved looks like this:
+	// | 2 1 0 0 | = | results[0] |
+	// | 1 4 1 0 | = | results[1] |
+	// | 0 1 4 1 | = | results[2] |
+	// | 0 0 1 2 | = | results[3] |
+	// Operations are carefully chosen to only modify the values in the
+	// diagonals and the results, leaving the 1s unchanged.
+	// All operations on diagonals are done by the solution cache, because the
+	// diagonals come out the same for all splines of a given size.
+	// Operation:  Add row[0] * -.5 to row[1] to zero [r1][c0].
+	results[1]-= results[0] * .5f;
+	for(size_t i= 1; i < last - 1; ++i)
+	{
+		// Operation:  Add row[i] / -[ri][ci] to row[i+1] to zero [ri+1][ci];
+		results[i]-= results[i-1] * (1.0f / diagonals[i]);
+	}
+	set_results(last, diagonals, results);
+}
+
+bool CubicSpline::check_minimum_size()
+{
+	size_t last= m_points.size();
+	if(last < 2)
+	{
+		m_points[0].b= m_points[0].c= m_points[0].d= 0.0f;
+		return true;
+	}
+	if(last == 2)
+	{
+		m_points[0].b= loop_space_difference(
+			m_points[1].a, m_points[0].a, m_spatial_extent);
+		m_points[0].c= m_points[0].d= 0.0f;
+		// These will be used in the looping case.
+		m_points[1].b= loop_space_difference(
+			m_points[0].a, m_points[1].a, m_spatial_extent);
+		m_points[1].c= m_points[1].d= 0.0f;
+		return true;
+	}
+	float a= m_points[0].a;
+	bool all_points_identical= true;
+	for(size_t i= 1; i < m_points.size(); ++i)
+	{
+		m_points[i].b= m_points[i].c= m_points[i].d= 0.0f;
+		if(m_points[i].a != a) { all_points_identical= false; }
+	}
+	return all_points_identical;
+}
+
+void CubicSpline::prep_inner(size_t last, vector<float>& results)
+{
+	for(size_t i= 1; i < last - 1; ++i)
+	{
+		results[i]= 3 * loop_space_difference(
+			m_points[i+1].a, m_points[i-1].a, m_spatial_extent);
+	}
+}
+
+void CubicSpline::set_results(size_t last, vector<float>& diagonals, vector<float>& results)
+{
+	// No more operations left, everything not a diagonal should be zero now.
+	for(size_t i= 0; i < last; ++i)
+	{
+		results[i]/= diagonals[i];
+	}
+	// Now we can go through and set the b, c, d values of each point.
+	// b, c, d values of the last point are not set because they are unused.
+	for(size_t i= 0; i < last; ++i)
+	{
+		size_t next= (i+1) % last;
+		float diff= loop_space_difference(
+			m_points[next].a, m_points[i].a, m_spatial_extent);
+		m_points[i].b= results[i];
+		m_points[i].c= (3 * diff) - (2 * results[i]) - results[next];
+		m_points[i].d= (2 * -diff) + results[i] + results[next];
+#define UNNAN(n) if(n != n) { n = 0.0f; }
```

Comment:
Would an (inline?) function be more appropriate than a define? Granted, it is undefined, but still.